src/mscorlib/src/System/Threading/SemaphoreSlim.cs

```diff
@@ -56,7 +56,13 @@ public class SemaphoreSlim : IDisposable
         // The number of synchronously waiting threads, it is set to zero in the constructor and increments before blocking the
         // threading and decrements it back after that. It is used as flag for the release call to know if there are
         // waiting threads in the monitor or not.
-        private volatile int m_waitCount;
```

Comment:
>volatile [](start = 16, length = 8)

why you removed the volatile on this field?

Reply:
It's always accessed from inside a lock (m_lockObj)

Reply:
Are we sure accessing it inside the lock is sufficient? My understanding is the lock implemented using managed thread Id with some compareexchange mechanism. so theoretically, there is a possibility the nonvolatile field will not get updated inside the lock block. I am not sure though :-)

Reply:
There's at least an interlocked operation at acquiring and releasing the lock, so it should get updated properly, otherwise it would be a bug in the lock. Another possibility is that volatile ensures some kind of ordering inside the lock, but I don't see anything like that, all of the state updates and most of the reads are inside the lock so changing ordering would not observe any difference in value of the fields.

Reply:
Here are a couple of enter/leave paths:
- https://github.com/dotnet/coreclr/blob/master/src/vm/syncblk.inl#L47
- https://github.com/dotnet/coreclr/blob/master/src/vm/syncblk.inl#L218

These are C++ versions of the assembly code for the fast paths that actually run, there are interlocked operations on enter and exit. Also checked all of the contention paths and wait paths, they also issue an interlocked operation.